### DISCUSSION TITLE
Custom Middleware: $request shows 127.0.0.1 "Symfony" as client, request() is right

### DISCUSSION THREAD

**Discussion by mschoeffmann** [livewire/livewire#5358] [7. Bugs (archived)]
Consider a custom middleware which checks the IP address of a session.

With the initial request (page-load), everything works fine.
But when the livewire lifecycle refreshes a component, the $request shows 127.0.0.1 as IP and "Symfony" as user agent.

I think it is very confusing to have different IP/Agent values, especially with a shared middleware wich is used for "normal" requests as well as livewire updates.

**Simplified Middleware**
```php
<?php

namespace App\Http\Middleware;

use Closure;
use Illuminate\Http\Request;

class ManagerIsAuthenticated
{
    public function handle(Request $request, Closure $next)
    {
        error_log('$request->ip() = '.$request->ip());
        error_log('request()->ip() = '.request()->ip());
        error_log('$request->userAgent() = '.$request->userAgent());
        error_log('request()->userAgent() = '.request()->userAgent());
        dump($request, request());

        if (request()->session()->has('manager') && request()->session()->get('manager') === request()->ip()) {
            return $next($request);
        }

        request()->session()->remove('manager');

        if (request()->header('x-livewire')) return abort(401);
        return redirect('/manager');
    }
}
```

**Simplified AppServiceProvider boot()**
```php
    public function boot()
    {
        Livewire::addPersistentMiddleware([
            ManagerIsAuthenticated::class,
        ]);
    }
```

**INITIAL OUTPUT (OK)**
```
laravel.test-1  |    WARN  $request->ip() = 172.22.0.1.  
laravel.test-1  |    WARN  request()->ip() = 172.22.0.1.  
laravel.test-1  |    WARN  $request->userAgent() = Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/107.0.0.0 Safari/537.36.  
laravel.test-1  |    WARN  request()->userAgent() = Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/107.0.0.0 Safari/537.36.  
```

**UPDATED OUTPUT (WRONG)**
```
laravel.test-1  |    WARN  $request->ip() = 127.0.0.1.  
laravel.test-1  |    WARN  request()->ip() = 172.22.0.1.  
laravel.test-1  |    WARN  $request->userAgent() = Symfony.  
laravel.test-1  |    WARN  request()->userAgent() = Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/107.0.0.0 Safari/537.36.  
```

And a screenshot of the $request dump. Looks like it is very simplified.

![dump-output](https://user-images.githubusercontent.com/9153512/201548319-8411b21a-0f6b-44ea-bb0f-3cb1e85bd254.png)

**Comment by joshhanley** [comment]
@mschoeffmann yeah that's because Livewire fakes the request on subsequent requests to ensure any middleware from the original route is applied correctly. But as we're not able to apply the middleware on the actual Livewire request (as they have already run), a fake request needs to be created. Hence why the middleware is receiving a localhost IP.

See here for where it happens https://github.com/livewire/livewire/blob/039c415f17878ed02a599acd0f1ac7c3d391493b/src/Controllers/HttpConnectionHandler.php#L30

Hope this helps!
